design-documents/Sale_Payment_Operation.md

```diff
@@ -0,0 +1,120 @@
+## Overview
+[Magento payment provider gateway](https://devdocs.magento.com/guides/v2.2/payments-integrations/bk-payments-integrations.html) doesn't support `sale` payment operation out of the box like `authorize` or `capture` operations.
+
+This operation is available for the majority of payment methods.
+
+Now, each payment method should implement an own workaround for `sale` payment operation, because OrderPayment calls capture instead sale:
+
+**Magento\Sales\Model\Order\Payment**
+```php
+protected function processAction($action, Order $order)
+{
+    $totalDue = $order->getTotalDue();
+    $baseTotalDue = $order->getBaseTotalDue();
+ 
+    switch ($action) {
+        ............................................................................
+        case \Magento\Payment\Model\Method\AbstractMethod::ACTION_AUTHORIZE_CAPTURE:
+            $this->setAmountAuthorized($totalDue);
+            $this->setBaseAmountAuthorized($baseTotalDue);
+            $this->capture(null);
+            break;
+        default:
+            break;
+    }
+}
+```
+[This article](https://community.magento.com/t5/Magento-DevBlog/The-Magento-Sale-Payment-Operation/ba-p/67251) describes how to solve this problem in each particular case but doesn't solve the problem in general.
+
+All payment integrations implement boilerplate code for `sale` operation:
+![](sale_payment_operation_img/sale_operation_flow.png)
+
+## Theory of Operation
+For the current moment, we have next relations in order payment workflow.
+![](sale_payment_operation_img/order_payment_current.png)
+
+I propose to introduce `SaleOperationInterface` and `Magento\Payment\Model\Method\Adapter` should implement this interface along with `Magento\Payment\Model\MethodInterface`.
```

Comment:
Please describe this interface with methods signatures. For sure it would be similar to canAuthorize/authorize, canCapture/capture methods but it's better to document this explicitly

Reply:
These methods are visible on diagrams https://github.com/magento-mpi/architecture/blob/b8380eee4248d17be578696f8a6d5af086888a37/design-documents/Sale_Payment_Operation.md, but I will add them in text content also

Reply:
On the diagram, there are no full methods signatures (input and output types)